doc.go

```diff
@@ -0,0 +1,107 @@
+// Generated by running
+//  go generate github.com/gonum/matrix
+// DO NOT EDIT.
+
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package matrix provides common error handling mechanisms for matrix operations
+// in mat64 and cmat128.
+//
+// Overview
+//
+// This section provides a quick overview of the matrix package. The following
+// sections provide more in depth commentary.
+//
+// matrix provides:
+//  - Error type definitions
+//  - Error recovery mechanisms
+//  - Common constants used by mat64 and cmat128
+//
+// Errors
+//
+// The mat64 and cmat128 matrix packages share a common set of errors
+// provided by matrix via the matrix.Error type.
+//
+// Errors are either returned directly or used as the parameter of a panic
+// depending on the class of error encountered. As is usually the case in
+// Go code, returned errors indicate that a call was not able to complete
+// successfully while panics generally indicate a programmer error or an
+// unrecoverable error.
+//
+// Examples of each type are found in the mat64 Solve methods, which find
+// x such that A*x = b.
+//
+// An error value is returned from the function or method when the operation
+// can meaningfully fail. The Solve operation is cannot complete if A is
+// singular. However, determining the singularity of A is most easily
+// discovered during the Solve procedure itself and is a valid result from
+// the operation, so in this case an error is returned.
+//
+// A function will panic when the input parameters are inappropriate for
+// the function. In Solve, for example, the number of rows of each input
+// matrix must be equal because of the rules of matrix multiplication.
+// Similarly, for solving A*x = b, a non-zero receiver must have the same
+// number of rows as A has columns and must have the same number of columns
+// as b. In all cases where a function will panic, conditions that would
```

Comment:
I don't think this is the basis - it puts the cart before the horse. We are mimicking a type system (we are actually implementing a type system), just one that is determined at run time. We preclude certain sets of behaviours by prohibiting some user actions. This is the foremost goal. The corollary of that prohibition is that we should (and do) make it possible for the user to determine whether they would be breaking the rules (if, for example, they were given untrusted input). The wording here reflects that priority. The converse, e.g. trying to solve a system which has a singular input, is a perfectly legitimate thing to do in the rules of linear algebra, it just fails. This is the difference IMO.